### PR TITLE
crypto: verify sig should be more specific

### LINKED ISSUE #52: `PublicKeySignatureVerifier::verify_signature` input signature type is too strict
**Describe the bug**

There exists two representations for Ed25519 signatures:

- `Ed25519Signature`
- `Signature`

The same would be true for Secp256k1 and P256 if not for #51.

But `PublicKeySignatureVerifier::verify_signature` can only accept a single type for the `signature` argument:

https://github.com/trilitech/tezedge/blob/6675c1cc23c07194ee5c1d1f965ddeaed8bd382b/crypto/src/lib.rs#L49-L58

Furthermore, `impl` for `PublicKeyEd25519` sets signature type to `Signature`:

https://github.com/trilitech/tezedge/blob/6675c1cc23c07194ee5c1d1f965ddeaed8bd382b/crypto/src/hash.rs#L651-L656

This requires awkward conversion from `Ed25519Signature` to `Signature`, something like this perhaps:

```rust
pk.verify_signature(Signature(ed25519sig.0), msg)
```

It would be more ergonomic to have a trait for signatures and accept that in `verify_signature` instead, or at least have an instance like this:

```rust
impl From<Ed25519Signature> for Signature {
    fn from(x: Ed25519Signature) -> Self {
        Signature(x.0)
    }
}
```

Ditto for Secp256k1 and P256 after #51 is addressed.

### PR DESCRIPTION
**Pull request by emturner**
we update `verify_sig` impls to require the specific signature type for the key being used.

To ease use with `Signature` enum, we provide `TryFrom<Signature> for <SigType>` impls.

Closes #52 

*NB* this does slightly the opposite of what was requested in #52. It's addressed, however, as it is much easier now to convert too the correct signature type, to pass through.

The corresponding `PublicKey` type in the SDK, will be able to use this to greater effect, and can indeed receive the more general `Signature` type.

This layout/structure is more in-line with the `signature` module provided in `octez/tezos_crypto`.

### CODE REVIEW

File: crypto/src/hash.rs [hunk 1]
```diff
@@ -1053,7 +1097,7 @@ mod tests {
         .unwrap();
         let sig = Signature::from_base58_check(
```
Comment (vapourismo) [inline crypto/src/hash.rs:1098]:
```suggestion
        let sig = Ed25519Signature::from_base58_check(
```

Why not use the more precise signature type directly?

File: crypto/src/signature.rs [hunk 2]
```diff
@@ -125,6 +136,71 @@ impl ::std::fmt::Display for Signature {
...
+impl TryFrom<Signature> for UnknownSignature {
+    type Error = TryFromSignatureError;
+
+    fn try_from(s: Signature) -> Result<Self, Self::Error> {
+        match s {
+            Signature::Ed25519(s) => Ok(s.into()),
+            Signature::Secp256k1(s) => Ok(s.into()),
+            Signature::P256(s) => Ok(s.into()),
+            Signature::Unknown(s) => Ok(s),
+            s => Err(Self::Error::InvalidKind(s.hash_type())),
```
Comment (vapourismo) [inline crypto/src/signature.rs:199]: BLS doesn't work?

Comment (emturner) [reply crypto/src/signature.rs:199]: Yeah, see here in [octez](https://gitlab.com/tezos/tezos/-/blob/master/src/lib_crypto/signature_v1.ml?ref_type=heads#L571)

TLDR; the signature length for all signatures is 64 bytes, *except for bls*, which is 96. Therefore everything can be a valid `unknown`, except for BLS sigs

File: crypto/src/hash.rs (hunk 1 above)
Comment (emturner) [reply crypto/src/hash.rs:1098]: The base58 check is specifically for the `Signature` / `UnknownSignature` encoding - so this will be parsed as a `Signature::Unknown` variant, and then convert it.

I could change this to use `UnknownSignature` directly though?

File: crypto/src/signature.rs (hunk 2 above)
Comment (vapourismo) [reply crypto/src/signature.rs:199]: Ok, I understand. But this is very confusing from a UX perspective. E.g. why does an unknown signature have a fixed length? I think this abstraction needs to be re-thought.

**Review by vapourismo — APPROVED** [review]

File: crypto/src/signature.rs (hunk 2 above)
Comment (emturner) [reply crypto/src/signature.rs:199]: I do kind of agree - it's definitely worth thinking about for future, especially if we can improve things on the octez side too.
